src/Stratis.Bitcoin.Features.Wallet/WalletRPCController.cs

```diff
@@ -321,6 +328,206 @@ public async Task<GetTransactionModel> GetTransactionAsync(string txid)
+            var setMap = new Dictionary<BitcoinPubKeyAddress, List<BitcoinPubKeyAddress>>();
+            foreach (var group in groupings)
+            {
+                var hits = new List<List<BitcoinPubKeyAddress>>();
+                foreach (var address in group)
+                {
+                    var mappedEntry = setMap.FirstOrDefault(m => m.Key == address);
+                    if (mappedEntry.Value != null)
+                        hits.Add(mappedEntry.Value);
+
+                    var merged = new List<BitcoinPubKeyAddress>(group);
+                    foreach (var hit in hits)
+                    {
+                        merged.AddRange(hit);
+                        uniqueGroupings.Remove(hit);
+                    }
+
+                    uniqueGroupings.Add(merged);
+
+                    foreach (var item in merged)
+                    {
+                        setMap[item] = merged;
+                    }
+                }
+            }
+
+            return uniqueGroupings;
+        }
+
+        private BitcoinPubKeyAddress GetBase58AddressFromScript(Script script)
+        {
+            PubKey payToPubKey = PayToPubkeyTemplate.Instance.ExtractScriptPubKeyParameters(script);
+
+            uint160 address = null;
+
+            if (payToPubKey != null)
+                address = new uint160(payToPubKey.Hash.ToBytes());
+
+            if (PayToPubkeyHashTemplate.Instance.CheckScriptPubKey(script))
+                address = new uint160(PayToPubkeyHashTemplate.Instance.ExtractScriptPubKeyParameters(script).ToBytes());
+
+            if (address != null)
+                return new BitcoinPubKeyAddress(new KeyId(address), this.Network);
+
+            return null;
+        }
+
+        /// <summary>
+        /// This will check the wallet's list of <see cref="HdAccount.InternalAddress"/>es to see if this address is
+        /// an address that received change.
+        /// </summary>
+        /// <param name="wallet">The wallet to use.</param>
+        /// <param name="coinType">The coin type associated with the wallet; Stratis or Bitcoin.</param>
+        /// <param name="txOutAddressBase58">The base58 address to verify from the <see cref="TxOut"/>.</param>
+        /// <returns><c>true</c> if the <paramref name="txOutAddressBase58"/> is a change address.</returns>
+        private bool IsChange(Wallet wallet, CoinType coinType, BitcoinPubKeyAddress txOutAddressBase58)
+        {
+            var internalAddresses = wallet.GetAccountsByCoinType(coinType).SelectMany(a => a.InternalAddresses);
+            var result = internalAddresses.FirstOrDefault(ia => ia.Address == txOutAddressBase58.ToString());
+            return result != null;
+        }
+
+        private async Task<bool> IsTxInMineAsync(Wallet wallet, CoinType coinType, IEnumerable<TransactionData> transactions, TxIn txIn)
+        {
+            var previousTransaction = transactions.FirstOrDefault(t => t.Id == txIn.PrevOut.Hash);
```

Comment:
If there are a very large number of transactions in the wallet, this is going to get called a lot. Maybe we should pass a dictionary to allow rapid lookup by txid instead of a list

Reply:
OK I've created a dictionary for faster lookups.